python-checks/src/main/java/org/sonar/python/checks/DjangoNonDictSerializationCheck.java

```diff
@@ -0,0 +1,114 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.checks;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import javax.annotation.CheckForNull;
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.symbols.Usage;
+import org.sonar.plugins.python.api.tree.Argument;
+import org.sonar.plugins.python.api.tree.AssignmentStatement;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.RegularArgument;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key = "S6560")
+public class DjangoNonDictSerializationCheck extends PythonSubscriptionCheck {
+
+  private static final String MESSAGE = "Use a dictionary object here, or set the \"safe\" flag to False.";
+  private static final String JSON_RESPONSE_FUNCTION_NAME = "django.http.JsonResponse";
+
+  private static final int MAX_RECURSION = 5;
+
+  @Override
+  public void initialize(Context context) {
+
+    context.registerSyntaxNodeConsumer(Tree.Kind.CALL_EXPR, ctx -> {
+      CallExpression callExpression = (CallExpression) ctx.syntaxNode();
+      Symbol symbol = callExpression.calleeSymbol();
+      if (symbol != null && JSON_RESPONSE_FUNCTION_NAME.equals(symbol.fullyQualifiedName())) {
+        checkForDictSerialization(ctx, callExpression);
+      }
+    });
+  }
+
+  private static void checkForDictSerialization(SubscriptionContext ctx, CallExpression callExpression) {
+    RegularArgument safe = TreeUtils.nthArgumentOrKeyword(2, "safe", callExpression.arguments());
+    if (safe == null || (safe.expression().is(Tree.Kind.NAME) && "True".equals(((Name) safe.expression()).name()))) {
+      RegularArgument dataArg = getFirstArgument(callExpression.arguments());
+      if (dataArg != null
+        && (!dataArg.expression().is(Tree.Kind.DICTIONARY_LITERAL))
+        && (!dataArg.expression().is(Tree.Kind.CALL_EXPR)
+          && !isDictAssignedToExpression(dataArg.expression()))) {
+        ctx.addIssue(dataArg, MESSAGE);
+      }
```

Comment:
Could we use type inference here instead, searching for type compatibility with dictionary types? (Might be prone to FPs though - I don't mind the current approach, this is just a suggestion).

That said, what if the data argument is a qualified expression like `my_obj.dict_field`? I have the feeling `isDictAssignedToExpression` would return false in that case, and we could raise a false positive. We'd probably want to have a test with this.
We also probably want to ensure we don't raise FPs with dictionary comprehensions.

In general, when in doubt, we should err on the side of not raising an issue, so maybe `isDictAssignedToExpression` could be renamed `couldExpressionBeADict` and return true whenever we don't know (so that if we forgot to consider an AST type, we don't raise FPs for it - and this would be consistent with the other method `couldDictBeAssignedToDataArg`).